File: Loenn/entities/PopRefill.lua
local drawableSprite = require("structs.drawable_sprite")

local FemtoHelperPopRefill = {}

FemtoHelperPopRefill.name = "FemtoHelper/PopRefill"
FemtoHelperPopRefill.depth = -100

FemtoHelperPopRefill.placements = {
    {
        name = "PopRefill",
        data = {
            oneUse = false,
            twoDash = false,
            spawnTime = 2.5,
            respawnTimer = 2.5,
            visualOffsetX = 0,
            visualOffsetY = 0
        }
    },
    {
        name = "PopRefillTwo",
        data = {
            oneUse = false,
            twoDash = true,
            spawnTime = 2.5,
            respawnTimer = 2.5,
            visualOffsetX = 0,
            visualOffsetY = 0
        }
    }
}

function FemtoHelperPopRefill.sprite(room, entity)
    local sprPath = entity.twoDash and "objects/refillTwo/" or "objects/refill/"

    local sprites = {}

    local vx = entity.visualOffsetX or 0;
    local vy = entity.visualOffsetY or 0;
    
    if vx ~= 0 or vy ~= 0 then
        local main = drawableSprite.fromTexture(sprPath.."outline", entity)
        table.insert(sprites, main)

        local offset_spr = drawableSprite.fromTexture(sprPath.."outline", entity)
        offset_spr:setColor({1, 1, 1, 0.5})
        offset_spr:addPosition(vx, vy)
        table.insert(sprites, offset_spr)
    else
        local main = drawableSprite.fromTexture(sprPath.."outline", entity)
        table.insert(sprites, main)
    end

    return sprites
end

return FemtoHelperPopRefill

File: Source/Entities/PopRefill.cs


// Celeste, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null
// Celeste.Refill
using System;
using System.Collections;
using static MonoMod.InlineRT.MonoModRule;

namespace Celeste.Mod.FemtoHelper.Entities;

[CustomEntity("FemtoHelper/PopRefill")]
public class PopRefill : CustomRefill
{

    public bool DoRespawn = false;
    public float SpawnTime = 2.5f;
    
    public PopRefill(EntityData data, Vector2 offset)
        : base(data.Position + offset, data.Bool("twoDash"), data.Bool("oneUse"))
    {
        RespawnTime = data.Float("respawnTime", 2.5f);
        SpawnTime = data.Float("spawnTime", 2.5f);
        this.SetRespawnGate(() => DoRespawn)
            .SetOnRespawn(() => outline.Color = Color.White);

        VisualOffset = data.Vector2("visualOffsetX", "visualOffsetY", Vector2.Zero);
    }
    
    public override void Added(Scene scene)
    {
        base.Added(scene);
        Collidable = false;
        sprite.Visible = false;
        flash.Visible = false;
        outline.Visible = true;
        Depth = 8999;
        respawnTimer = SpawnTime;
        Add(new Coroutine(StartRoutine()));
    }

    public IEnumerator StartRoutine()
    {
        while (true)
        {
            if (CollideCheck<Player>())
            {
                Audio.Play("event:/game/03_resort/fluff_tendril_touch", Position);
                outline.Color *= 0.2f;
                break;
            }
            yield return null;
        }
        while (true)
        {
            if (!CollideCheck<Player>())
            {
                Audio.Play("event:/game/03_resort/fluff_tendril_emerge", Position);
                outline.Color *= 2.5f;
                break;
            }
            yield return null;
        }
        DoRespawn = true;
        respawnTimer = SpawnTime;
    }
}
